nfs/nfs.go

```diff
@@ -51,19 +55,37 @@ type Sattr3 struct {
 	Mode  SetMode
 	UID   SetUID
 	GID   SetUID
-	Size  uint64
-	Atime NFS3Time
-	Mtime NFS3Time
+	Size  SetSize
+	Atime SetTime
+	Mtime SetTime
 }
 
 type SetMode struct {
-	Set  uint32
-	Mode uint32
+	SetIt bool   `xdr:"union"`
+	Mode  uint32 `xdr:"unioncase=1"`
 }
```

Comment:
Sorry for being late to the game, but all sequences of "isset bool + value" are described in the RFC as optional data type. This type supported using `xdr:"optional"` and a pointer type, so basically you can replace `Mode SetMode` with ``Mode *uint32 `xdr:"optional"` `` and then check `Mode != nil` to find out when `Mode` is set.

Reply:
@rasky Thanks for the review at all! This was a very targeted PR done with the NFS RFC in one hand, wireshark trace in the other. Does the same `xdr:optional` and pointer values apply for the multi-value unions as well, e.g. [this struct](https://github.com/vmware/go-nfs-client/pull/2/files/5173af60f8ac8b0b15900e14811074cea7b5ae6e#diff-6caff2ea91b64ba7c7b9ce3537406382R198)?

@matthewavery If you know wireshark/delve can you pick this up and use it to teach those tools to @AngieCris and/or @anchal-agrawal. If not I'll run that with the three of you next time I'm in Austin.

Reply:
@hickeng the PR is technically correct, I'm just suggesting a better usage of the Go XDR library. If you read [RFC4506](https://tools.ietf.org/html/rfc4506), there are two different data types: discriminated unions (§4.15) and optional types (§4.19). 

Optional types are defined as a special case of discriminated unions, where the discriminator is a boolean type, and the FALSE case is empty; so it does make sense that you can use a discriminated union for those structures, it's just that the optional type is more easy to reason, lets you write less code, and more readable.

NFS RFC has many optional types, but it's true that they're described as XDR unions in the spec. I'm not sure why, must be some historical reason, but the two are really equivalent.

As for applying optional values to multi-field structs, it's sufficient to use the correct level of indirection:

```go
type WccData struct {
	Before *struct {
		Size  uint64
		MTime NFS3Time
		CTime NFS3Time
	} `xdr:"optional"`
	After PostOpAttr
}
```

Notice that you need to change `Before` to be a pointer, in addition to adding the optional tag.